### PR TITLE
[SYCLomatic] Support asm  fence api

### PR DESCRIPTION
**Pull request by abhilash1910**
Proposed support for :
asm_volatile("fence.acq_rel;");

@yihanwg , @zhimingwang36

### CODE REVIEW

File: clang/test/dpct/asm/fence.cu [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+// RUN: %if build_lit %{icpx -c -fsycl %T/fence/fence.dp.cpp -o %T/fence/fence.dp.o %}
+
+// clang-format off
+// CHECK: #include <cmath>
+#include <cuda_runtime.h>
+#include <cstdint>
+
+__global__ void fence() {
+
+  // CHECK: sycl::group_barrier(item_ct1.get_group());
```
Comment (zhiweij1) [inline clang/test/dpct/asm/fence.cu]: Is `sycl::atomic_fence(sycl::memory_order::acq_rel, sycl:: memory_scope::device)` a better mapping?

Comment (abhilash1910) [reply clang/test/dpct/asm/fence.cu]: Yes @zhiweij1 I was also thinking to add atomic_fence, let me modify it. Thanks

File: clang/test/dpct/asm/fence.cu [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+// RUN: %if build_lit %{icpx -c -fsycl %T/fence/fence.dp.cpp -o %T/fence/fence.dp.o %}
+
+// clang-format off
+// CHECK: #include <cmath>
+#include <cuda_runtime.h>
+#include <cstdint>
+
+__global__ void fence() {
+
+  // CHECK: sycl::atpmic_fence(sycl::memory_order::acq_rel, sycl::memory_scope::device);
```
Comment (yihwang-nv) [inline clang/test/dpct/asm/fence.cu]: It's a typo? sycl::atomic_fence but not sycl::atpmic_fence.

File: clang/lib/DPCT/AsmMigration.cpp [hunk 3]
```diff
@@ -1987,6 +1987,15 @@ class SYCLGen : public SYCLGenBase {
     return SYCLGenSuccess();
   }
 
+  bool handle_fence(const InlineAsmInstruction *I) override {
+
```
Comment (yihwang-nv) [inline clang/lib/DPCT/AsmMigration.cpp]: Let's remove extra blank line.

**Review by yihwang-nv — CHANGES_REQUESTED** [review]
Let's  further add a test in https://github.com/oneapi-src/SYCLomatic-test, to make sure the behavior is correct.

**Comment by yihwang-nv** [comment]
Don't forget to update asm status in APINames_ASM.inc

File: clang/test/dpct/asm/fence.cu (hunk 2 above)
Comment (abhilash1910) [reply clang/test/dpct/asm/fence.cu]: Yes fixed

File: clang/lib/DPCT/APINames_ASM.inc [hunk 4]
```diff
@@ -67,7 +67,7 @@ ENTRY("dp4a", "dp4a", true, NO_FLAG, P1, "Successful")
 ENTRY("elect", "elect", false, NO_FLAG, P1, "Comment")
 ENTRY("ex2", "ex2", true, NO_FLAG, P1, "Successful")
 ENTRY("exit", "exit", false, NO_FLAG, P1, "Comment")
-ENTRY("fence", "fence", false, NO_FLAG, P1, "Comment")
+ENTRY("fence", "fence", true, NO_FLAG, P1, "Successful")
```
Comment (yihwang-nv) [inline clang/lib/DPCT/APINames_ASM.inc]: I think it should be partial, it's not fully supported.

File: clang/lib/DPCT/AsmMigration.cpp [hunk 5]
```diff
@@ -1987,6 +1987,14 @@ class SYCLGen : public SYCLGenBase {
     return SYCLGenSuccess();
   }
 
+  bool handle_fence(const InlineAsmInstruction *I) override {
```
Comment (yihwang-nv) [inline clang/lib/DPCT/AsmMigration.cpp]: Need check the instruction qualifiers.

File: clang/test/dpct/asm/fence.cu [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// UNSUPPORTED: cuda-8.0, cuda-9.0, cuda-9.1, cuda-9.2, cuda-10.0, cuda-10.1, cuda-10.2
+// UNSUPPORTED: v8.0, v9.0, v9.1, v9.2, v10.0, v10.1, v10.2
+// RUN: dpct --format-range=none -out-root %T/fence %s --cuda-include-path="%cuda-path/include" -- -std=c++14 -x cuda --cuda-host-only
+// RUN: FileCheck %s --match-full-lines --input-file %T/fence/fence.dp.cpp
+// RUN: %if build_lit %{icpx -c -fsycl %T/fence/fence.dp.cpp -o %T/fence/fence.dp.o %}
+
+// clang-format off
+// CHECK: #include <cmath>
```
Comment (yihwang-nv) [inline clang/test/dpct/asm/fence.cu]: Why check cmath here?

**Review by yihwang-nv — CHANGES_REQUESTED** [review]

File: clang/lib/DPCT/APINames_ASM.inc (hunk 4 above)
Comment (abhilash1910) [reply clang/lib/DPCT/APINames_ASM.inc]: changed

**Review by yihwang-nv — APPROVED** [review]
This approval only pokes CI, but not +1 for the PR.

**Comment by abhilash1910** [comment]
Closed as present

**Pull request closed without merging**